File: setops.py
import sys
import re

def read_files(file_path):

    # This is will make sure the things we pass in are files, if not an error  will be thrown 
    try:
        # This reads the file and returns the file.
        with open(file_path, 'r') as file:
            content = file.read()
            return file_path
    # If the file is not found, then the error is printed
    except FileNotFoundError:
        print('The file has not been found')

# We will check what the operation is and from there we will go to the function according to that 
def perform_operation(txt1, txt2, operation):

    # If difference operation is selected
    if operation == 'difference':
        difference_function(txt1, txt2)

    # If union operator is selected
    elif operation == 'union':
        union_function(txt1, txt2)

    # If intersection is selected intersection
    elif operation == 'intersection':
        intersection_function(txt1, txt2)

    else:
        print("sorry but that is not a valid operation.")
        sys.exit(1)


# This is for the difference operater function
def difference_function(txt1, txt2):

    # This creates two variables using the same function to store the the list of each file.
    list_1 = set(file_into_a_list(txt1))
    list_2 = set(file_into_a_list(txt2))

    # Creates a list that is sorted, and in difference
    difference_result = sorted(list(list_1.difference(list_2)))

    # This writes the word into the result txt file.
    # Essentially what this does is create file name result.txt and for every x in the length of difference_result, write that word into the new file with \newLine 
    # UNLESS it is the last word then you can just write the word without the new line 

    result_file = 'result.txt'
    with open(result_file, 'w') as result_file:
        for x in range(len(difference_result)):
            if x == len(difference_result) - 1:
                result_file.write(difference_result[x])
            else:
                result_file.write(difference_result[x]+ '\n')
            
    result_file.close()


# This is for the union operater function
def union_function(txt1, txt2):

    # This creates two variables using the same function to store the the list of each file.
    list_1 = set(file_into_a_list(txt1))
    list_2 = set(file_into_a_list(txt2))


    union_result = sorted(list(list_1.union(list_2)))
    
    # This writes the word into the result txt file.
    result_file = 'result.txt'
    with open(result_file, 'w') as result_file:
        for x in range(len(union_result)):
            if x == len(union_result) - 1:
                result_file.write(union_result[x])
            else:
                result_file.write(union_result[x]+ '\n')
            
    result_file.close()
    

# This is for the intersection operater function 
def intersection_function(txt1,txt2):

    # This creates two variables using the same function to store the the list of each file.
    list_1 = set(file_into_a_list(txt1))
    list_2 = set(file_into_a_list(txt2))

    print(list_1)
    print(list_2)

'''
    # This is where the actual intersection takes places but the value is in a set
    intersection_result = sorted(list(list_1.intersection(list_2)))
    
    # This writes the word into the result txt file.
    result_file = 'result.txt'
    with open(result_file, 'w') as result_file:
        for x in range(len(intersection_result)):
            if x == len(intersection_result) - 1:
                result_file.write(intersection_result[x])
            else:
                result_file.write(intersection_result[x]+ '\n')
            
    result_file.close()
'''


    
# This converts the txt file into a list.
def file_into_a_list(txt):
    try:
        with open(txt, 'r') as file:
            # Reads each line in the file, and withing each line it will read each word and split \.(?!d) -> all . that have zero to one occurences of no digits after it
            # |[\s,]+ -> OR split according to white space, and comma. And return that. ***
            words = [word.strip().lower() for line in file for word in re.split(r'\.(?!\d)|[\s,+-=/]+', line) if word]
            return words
    except FileNotFoundError:
        print(f"Error: File '{txt}' not found.")
        return []
    

    
    


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Format: python setops.py \"set1=[filename]; set2=[filename];operation=[difference|union|intersection]\"")
        sys.exit(1)

    # argv[1] is the string of set1, set2, and operation and its splits it with semi colon
    args = sys.argv[1].split(';')

    # There are three elements now which each one of them being assigned a variable.
    set1 = read_files(args[0].split('=')[1])
    set2 = read_files(args[1].split('=')[1])
    operation = args[2].split('=')[1]


    # If the correct parameters are passed in then start the actually part of the code.
    perform_operation(set1, set2, operation)

File: setops.js
const fs = require('fs');

function getWords(txt, callback) {
    let words = []
    fs.readFile(txt,'utf-8', (err,data) => {
        if (err) {
            console.error(err);
            callback(err, null);
        } else{
            callback(null,mergeSort(removeDuplicates(parseWords(words, data, 0), 0)));
        }
    });
        
}


function removeDuplicates(wordsList, index) {
    if (index + 1 > wordsList.length) {
        return wordsList;
    } else {
        return removeDuplicates(removeDupHelp(wordsList, index + 1, wordsList[index]), index + 1);
    }
}

function removeDupHelp(wordsList, index, target) {
    if (index > wordsList.length - 1) {
        return wordsList;
    } else if (wordsList[index] === target) {
        return removeDupHelp(wordsList.slice(0, index).concat(wordsList.slice(index + 1)), index, target);
    } else {
        return removeDupHelp(wordsList, index + 1, target);
    }
}

function intersection(wordsList1, wordsList2, index) {
    if (index >= wordsList1.length - 1) {
        return_result_file(wordsList1);
    } else {
        if (removeIntersectionHelp(wordsList2, 0, wordsList1[index])) {
            return intersection(wordsList1, wordsList2, index + 1);
        } else {
            return intersection(wordsList1.slice(0, index).concat(wordsList1.slice(index + 1)), wordsList2, index);
        }
    }
}

function removeIntersectionHelp(wordsList, index, target) {
    if (index > wordsList.length - 1) {
        return false;
    } else if (wordsList[index] === target) {
        return true;
    } else {
        return removeIntersectionHelp(wordsList, index + 1, target);
    }
}

function difference(wordsList1, wordsList2, index) {
    if (index >= wordsList1.length) {
        return_result_file(wordsList1);
    } else {
        if (removeDifferenceHelp(wordsList2, 0, wordsList1[index])) {
            return difference(wordsList1, wordsList2, index + 1);
        } else {
            return difference(wordsList1.slice(0, index).concat(wordsList1.slice(index + 1)), wordsList2, index);
        }
    }
}

function removeDifferenceHelp(wordsList, index, target) {
    if (index > wordsList.length - 1) {
        return true;
    } else if (wordsList[index] === target) {
        return false;
    } else {
        return removeDifferenceHelp(wordsList, index + 1, target);
    }
}

function union(wordsList1, wordsList2) {
    return_result_file(mergeSort(removeDuplicates(wordsList1.concat(wordsList2), 0)));
}


function parseWords(wordsList, line, index) {
    
    while (index < line.length) {
        if (/[0123456789]/.test(line[index])) {
            index += 1;
        } else if (/[a-z]/.test(line[index])) {
            index += 1;
        } else if (/[A-Z]/.test(line[index])) {
            line =
                line.slice(0, index) +
                String.fromCharCode(line.charCodeAt(index) + 32) +
                line.slice(index + 1);
            index += 1;
        } else {
            if (line[index] === ".") {
                if (index !== 0) {
                    if (index !== line.length - 1) {
                        if (/[0123456789]/.test(line[index - 1])) {
                            if (/[012345679]/.test(line[index + 1])) {
                                index += 1;
                            } else {
                                wordsList.push(line.slice(0, index));
                                line = line.slice(index + 1);
                                index = 0;
                            }
                        } else {
                            wordsList.push(line.slice(0, index));
                            line = line.slice(index + 1);
                            index = 0;
                        }
                    } else {
                        wordsList.push(line.slice(0, index));
                        line = line.slice(index + 1);
                        index = 0;
                    }
                } else {
                    line = line.slice(index + 1);
                    index = 0;
                }
            } else {
                if (index !== 0) {
                    wordsList.push(line.slice(0, index));
                }
                line = line.slice(index + 1);
                index = 0;
            }
        }
    }

    if (index !== 0) {
        wordsList.push(line);
    }

    return wordsList
}

function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    } else {
        return merged(mergeSort(arr.slice(0, Math.floor(arr.length / 2))), mergeSort(arr.slice(Math.floor(arr.length / 2))), 0, 0);
    }
}


function merged(left, right, i, j) {
    let mergedArray = [];
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            mergedArray.push(left[i]);
            i++;
        } else {
            mergedArray.push(right[j]);
            j++;
        }
    }
    mergedArray.push(...left.slice(i), ...right.slice(j));
    return mergedArray;
}


function perform_operation(txt1, txt2, operation) {
    getWords(txt1, (err, words1) => {
        if (err) {
            console.error(`Error reading file ${txt1}: ${err.message}`);
            process.exit(1);
        }

        getWords(txt2, (err, words2) => {
            if (err) {
                console.error(`Error reading file ${txt2}: ${err.message}`);
                process.exit(1);
            }

            if (operation === 'difference') {
                difference((words1),(words2), 0);
            } else if (operation === 'union') {
                union((words1),(words2));
            } else if (operation === 'intersection') {
                intersection((words1), (words2), 0);
            } else {
                console.log("Sorry, but that is not a valid operation.");
                process.exit(1);
            }

        });
    });
}



function return_result_file(txtList) {
    fs.writeFileSync('result.txt', txtList.join('\n'), 'utf-8');
}


if (require.main === module) {
    if (process.argv.length !== 3) {
        console.log("Format: node setops.js \"set1=[filename]; set2=[filename];operation=[difference|union|intersection]\"");
        process.exit(1);
    }

    const args = process.argv[2].split(';');


    perform_operation((args[0].split('=')[1]), (args[1].split('=')[1]), args[2].split('=')[1]);
}
